### PR TITLE
Extensible name section support

### PR DESCRIPTION
**Pull request by pipcet**
This PR should change the code to support the new extensible name section rather than the previous now-obsolete layout.

However, I've so far been unable to find any code (except for my own) that produces or reads the new format, so it's not as well-tested as it should be; it might be worth waiting for adoption in other tools, in any case.

See https://github.com/WebAssembly/wabt/issues/334.

### CODE REVIEW

**Comment by pipcet** [comment]
@binji:
> We'll want to support the other name sections for wasm2wast (binary-reader-ast) and wasmdump (binary-reader-objdump), but that seems reasonable to do in a separate PR.

I'm not sure I understand; are you saying you want to keep support for the old pre-MVP format?

**Comment by pipcet** [comment]
Oops, thought it was running the tests locally, but it wasn't. I'll fix those.

**Comment by binji** [comment]
> I'm not sure I understand; are you saying you want to keep support for the old pre-MVP format?

Sorry, I said we want to support the other name sections but I meant we want to support reading the new kinds of names in those files (e.g. names of labels, globals, memories, etc.)

**Comment by pipcet** [comment]
> I meant we want to support reading the new kinds of names in those files (e.g. names of labels, globals, memories, etc.)

I may have missed something (again), but the spec doesn't mention those yet. I'm in favor of adding those as soon as they appear in the spec.

(There are quite a few commits now, so let me know if and when you want to merge so I can squash them into a single commit to keep the history readable.)

**Comment by binji** [comment]
You're right, I assumed they had been added since that was the purpose of restructuring the name section. :)

File: src/binary-writer.cc [hunk 1]
```diff
@@ -300,6 +304,26 @@ static void end_section(Context* ctx) {
   ctx->last_section_leb_size_guess = 0;
 }
 
+static void begin_subsection(Context* ctx,
+                             const char* name,
+                             size_t leb_size_guess) {
+  assert(ctx->last_subsection_leb_size_guess == 0);
+  char desc[100];
+  snprintf(desc, sizeof(desc), "subsection \"%s\"", name);
```
Comment (binji) [inline src/binary-writer.cc]: use wabt_snprintf instead

**Comment by pipcet** [comment]
Thanks! I've addressed those two, do let me know if you spot anything else.

I've joined the CG now.

**Comment by binji** [comment]
Great!

**Pull request merged**